bin/spack-modify-makefile-push-buildcache

```diff
@@ -3,14 +3,57 @@
 import fileinput
 import re
 import sys
+import subprocess
 
+# In order to speed up building packages with Spack, we use a buildcache. To further optimize the process, we want the following behavior:
+# - Install from buildcache
+# - If installation from buildcache was not successfull (the package is not in the cache or it is corrupted), build from source
+# - Only if the package was built from source, push to the buildcache. Moreover, if pushing to the buildcache fails, it is not fatal.
+
+# While Spack is able to install from buildcache and build from source as a fallback, it can only push to the buildcache no matter which way the package was installed.
+# As this is going to be executed in a CI pipeline, we can expect most of the packeges are going to be installed from the buildcache. This means most of the packages would be first pulled and then pushed to the buildcache without changing anything, slowing down the whole process.
+
+# To achieve the expected behavior, we implement it by modifying the "spack install" call in the makefile generated by Spack.
+# We first install from the build cache and if it fails, only then we compile the package and push it to the buildcache.
+
+# This script replaces "spack install" call in a makefile, one of these
+#
 # Spack-0.18
+# TODO
 #+$(SPACK) -e '/opt/spack-environment' install $(SPACK_INSTALL_FLAGS) --only-concrete --only=package --no-add /$(notdir $@) && touch $@
+#
 # Spack 0.19
 #+$(SPACK) -e '/opt/spack-environment' install $(SPACK_BUILDCACHE_FLAG) $(SPACK_INSTALL_FLAGS) --only-concrete --only=package --no-add /$(notdir $@) # $(SPEC)
+#
+# Spack 0.20
+#+$(SPACK) -e '/opt/spack-environment' install $(SPACK_BUILDCACHE_FLAG) $(SPACK_INSTALL_FLAGS) --only-concrete --only=package /$(HASH) # $(SPEC)
+
+# with something like this
+#
+# Spack 0.19
+# +$(SPACK) -e '/opt/spack-environment' install --cache-only $(SPACK_INSTALL_FLAGS) --only-concrete --only=package --no-add /$(notdir $@) || \
+#   ( $(SPACK) -e '/opt/spack-environment' install --no-cache $(SPACK_INSTALL_FLAGS) --only-concrete --only=package --no-add /$(notdir $@) && \
+#     ( $(SPACK) -e '/opt/spack-environment' buildcache create --allow-root --rebuild-index --force -m s3_cscs_buildcache --only=package /$(notdir $@) || true ) ) # $(SPEC)
+#
+# Spack 0.20
+# +$(SPACK) -e '/opt/spack-environment' install --cache-only $(SPACK_INSTALL_FLAGS) --only-concrete --only=package /$(HASH) || \
+#   ( $(SPACK) -e '/opt/spack-environment' install --no-cache $(SPACK_INSTALL_FLAGS) --only-concrete --only=package /$(HASH) && \
+#     ( $(SPACK) -e '/opt/spack-environment' buildcache create --allow-root --rebuild-index --force -m s3_cscs_buildcache --only=package /$(notdir $@) || true ) ) # $(SPEC)
```

Comment:
This should be for version 0.20 `/$(HASH)` instead of `/$(notdir $@)`, The current push to buildcache (`buildcache create`), is currently hardcoded a few lines below. This should be also dependent on the version (or find a way to make it generic that in 0.19 it is `/$/(notdir $@)` and in 0.20 `/$(HASH)`